### PR TITLE
Add Neural Network Regression example on BodyFat.tsv

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hi everyone, I think that mlpack/mlpack#1529 might be better suited here. So I have made some changes in the code (that works for the current codebase). @zsogitbe, @zoq Let me know if this makes sense here. I will add the tutorial and a notebook here as well if it's okay with you. Sorry if I overstepped my bounds, then I can take this down. Looking forward to hearing from you. 
Thanks a lot.

### CODE REVIEW

**Comment by birm** [comment]
I do not think adding an example to the examples repo is overstepping bounds :wink: 
I think the notebooks are pretty cool, and I agree that'd be a good addition. I've been meaning to try xeus myself.

**Comment by kartikdutt18** [comment]
Great.

>  I've been meaning to try xeus myself.

I think mlpack-lab is amazing, much easier to use than setting up xeus cling locally.

**Comment by kartikdutt18** [comment]
@mlpack-jenkins test this please

**Comment by kartikdutt18** [comment]
@mlpack-jenkins test this please

**Comment by kartikdutt18** [comment]
@mlpack-jenkins test this please

**Comment by kartikdutt18** [comment]
@mlpack-jenkins test this please

**Comment by zsogitbe** [comment]
Did you change something? Did you test the results?

**Comment by kartikdutt18** [comment]
> Did you change something? Did you test the results?

Hi @zsogitbe, I did change a few things. They are listed below:
1. Use internal scaler of mlpack.
2. Simplify example by using ensmallen callbacks.
3. Simplify the example in terms of comments.

I'll push the results in form of a notebook tomorrow.

**Comment by kartikdutt18** [comment]
Here are the last few lines from after execution : 
```
Epoch 59/8513
0.114143============] 100% - ETA: 0s - loss: 0.0899672
8/8 [============] 100% - 0s 0ms/step - loss: 0.114143
Finished training. 
 Saving Model
Model saved in nn_regressor.bin
Mean Squared Error on Prediction data points: 0.312585
```

**Comment by zsogitbe** [comment]
> Here are the last few lines from after execution :
> ```
> Epoch 59/8513
> 0.114143============] 100% - ETA: 0s - loss: 0.0899672
> 8/8 [============] 100% - 0s 0ms/step - loss: 0.114143
> Mean Squared Error on Prediction data points: 0.312585
> ```
Hi Kartik,
The results are not very clear to me. What are those loss values? Is the iteration stopped at epoch 59? The MSE of 0.31 for the bodyfat example is very high even for an example. The MSE should be around 0.008 or lower.
Br,
Zoltan

**Comment by kartikdutt18** [comment]
Hey Zoltan,
Thanks for getting back so quickly. Sorry, I'll clarify here.

> Is the iteration stopped at epoch 59?

Yes, I have used early stopping on training loss. So the callback stopped the prediction at the 59th epoch.

> The MSE of 0.31 for the bodyfat example is very high even for an example. The MSE should be around 0.008 or lower.

I agree, I'll try tweaking the model further to reduce it more. 

> What I would like to see is the MSE for the training and also for the test set.

The training MSE at 59th epoch is :  0.0899672
The validation MSE is : 0.31

I'll try out some changes now and hopefully that will improve the performance.

Thanks a lot.
Regards,
Kartik.

**Comment by kartikdutt18** [comment]
Hey @zoltan, I have made some more changes here are the updated results : 
The model ran for 100 epochs, 
Training loss after 100 epochs : 3.01048e-05

Validation loss after 100 epochs : 0.000464804

I am also pasting the output log below : 
```
Epoch 100/2838
3.01048e-05==============================================================================================] 100% - ETA: 0s - loss: 2.37284e-05
8/8 [====================================================================================================] 100% - 0s 0ms/step - loss: 3.01048e-05
Finished training. 
 Saving Model
Model saved in nn_regressor.bin
Mean Squared Error on Prediction data points: 0.000464804
```

Thanks a lot.

Regards.

**Comment by kartikdutt18** [comment]
Hey Zoltan,
I ran the original, final here is the output log :
It ran for 80 epochs with mean squared error in both datasets shown below.

```
80 - accuracy: train = 0.000880932 (MSQE), valid = 0.13525 (MSQE)
```

I did have to make some changes to accommodates changes in mlpack such as use ensmallen and change saving of files.

Kindly let me know what you think. Thanks.
Regards,
Kartik.

**Comment by zsogitbe** [comment]
> Hey @zoltan, I have made some more changes here are the updated results :
> The model ran for 100 epochs,
> Training loss after 100 epochs : 3.01048e-05
> Validation loss after 100 epochs : 0.000464804
> Thanks a lot.
> Regards.

That is good! Good work!

**Comment by kartikdutt18** [comment]
> That is good! Good work!

Thanks a lot. I will update the note book to show the same. Thanks for getting back so quickly.

Regards,
Kartik.

**Comment by kartikdutt18** [comment]
Hey everyone, I have updated the notebook as well. I have used lesser number of epochs to keep the notebook a bit short (added a comment for the same).

Here are results from the notebook :

Epochs : 30
Training MSE : 0.000323442
Validation MSE : 0.0087528

Thanks a lot,
Regards,
Kartik.

**Comment by kartikdutt18** [comment]
Thanks @birm, for the review.

**Comment by kartikdutt18** [comment]
Thanks @birm, @zsogitbe for all the help and the reviews. Thanks a lot.
Regards.
